index.js

```diff
@@ -430,35 +431,47 @@ module.exports = class Hypercore extends EventEmitter {
   }
 
   replicate (isInitiator, opts = {}) {
+    opts = { firewall: this.firewall, ...opts }
     // Only limitation here is that ondiscoverykey doesn't work atm when passing a muxer directly,
     // because it doesn't really make a lot of sense.
-    if (Protomux.isProtomux(isInitiator)) return this._attachToMuxer(isInitiator, opts)
+    if (Protomux.isProtomux(isInitiator)) return this._attachToMuxer(isInitiator, true, opts)
 
     const protocolStream = Hypercore.createProtocolStream(isInitiator, opts)
     const noiseStream = protocolStream.noiseStream
     const protocol = noiseStream.userData
     const useSession = !!opts.session
 
-    this._attachToMuxer(protocol, useSession)
+    this._attachToMuxer(protocol, useSession, opts)
 
     return protocolStream
   }
 
-  _attachToMuxer (mux, useSession) {
+  _attachToMuxer (mux, useSession, opts) {
     if (this.opened) {
-      this._attachToMuxerOpened(mux, useSession)
+      this._attachToMuxerOpened(mux, useSession, opts)
     } else {
-      this.opening.then(this._attachToMuxerOpened.bind(this, mux, useSession), mux.destroy.bind(mux))
+      this.opening.then(this._attachToMuxerOpened.bind(this, mux, useSession, opts), mux.destroy.bind(mux))
     }
 
     return mux
   }
 
-  _attachToMuxerOpened (mux, useSession) {
+  _attachToMuxerOpened (mux, useSession, opts = {}) {
     // If the user wants to, we can make this replication run in a session
     // that way the core wont close "under them" during replication
-    const session = useSession ? this.session() : null
-    this.replicator.attachTo(mux, session)
+
+    const maybeAttach = (allow) => {
+      if (!allow) return mux.destroy()
```

Comment:
@mafintosh I don't think this is correct. Suppose someone passes `replicate` a `Protomux` instance, should we really destroy if replication is firewalled? What if they are running other protocols on the mux instance? It would be nice to signal firewalling a bit more responsibly. If we noop, the caller has no indication that replication has been firewalled. We can't throw because the chain of methods is sync and throwing is ugly, no need to panic on getting firewalled. We could thread `protocolStream` through from `replicate` and emit a `firewalled` event...that's the best I can come up with right now. Some guidance would be appreciated.